File: examples/typescript/idp-client.ts
/**
 * Example TypeScript client for IDP: fetch every config/secret for a
 * service+environment (decrypted into an object), and every feature flag
 * (into an object of name -> enabled). Both use the same service client API key.
 *
 *   npm install fernet
 *   IDP_API_KEY=... IDP_ENCRYPTION_KEY=... npx tsx idp-client.ts
 *
 * Requires Node 18+ (global fetch).
 */
import Fernet from "fernet";

interface ConfigResponse {
  id: string;
  service: string;
  environment: string;
  key: string;
  value: string; // Fernet-encrypted with the calling client's encryption_key
  is_secret: boolean;
  type: "boolean" | "string" | "number" | "object" | "array";
}

interface FeatureFlagResponse {
  id: string;
  service: string;
  environment: string;
  name: string;
  description: string;
  is_enabled: boolean;
}

/**
 * Fetches every config/secret for a service+environment via the S2S API,
 * decrypts each value with this client's encryption_key, and returns them
 * as a plain { KEY: value } object.
 */
export async function getDecryptedConfigs(
  baseUrl: string,
  apiKey: string,        // "<key_id>.<secret>" from POST /api/v1/auth/s2s/clients
  encryptionKey: string, // this client's Fernet key, from the same response
  service: string,
  environment: string
): Promise<Record<string, string>> {
  const url = `${baseUrl}/api/v1/config/configs/list?service=${encodeURIComponent(
    service
  )}&environment=${encodeURIComponent(environment)}`;

  const res = await fetch(url, {
    headers: { "X-API-Key": apiKey },
  });
  if (!res.ok) {
    throw new Error(`Failed to list configs: ${res.status} ${await res.text()}`);
  }

  const configs: ConfigResponse[] = await res.json();
  const secret = new Fernet.Secret(encryptionKey);

  const result: Record<string, string> = {};
  for (const config of configs) {
    const token = new Fernet.Token({ secret, token: config.value, ttl: 0 });
    result[config.key] = token.decode();
  }
  return result;
}

/**
 * Fetches every feature flag for a service+environment and returns them as
 * a plain { FLAG_NAME: enabled } object. Flags aren't encrypted.
 */
export async function getFeatureFlags(
  baseUrl: string,
  apiKey: string,
  service: string,
  environment: string
): Promise<Record<string, boolean>> {
  const url = `${baseUrl}/api/v1/config/feature-flags?service=${encodeURIComponent(
    service
  )}&environment=${encodeURIComponent(environment)}`;

  const res = await fetch(url, {
    headers: { "X-API-Key": apiKey },
  });
  if (!res.ok) {
    throw new Error(`Failed to list feature flags: ${res.status} ${await res.text()}`);
  }

  const flags: FeatureFlagResponse[] = await res.json();

  const result: Record<string, boolean> = {};
  for (const flag of flags) {
    result[flag.name] = flag.is_enabled;
  }
  return result;
}

async function main() {
  const baseUrl = process.env.IDP_BASE_URL || "http://localhost:8000";
  const apiKey = process.env.IDP_API_KEY!;

  const configs = await getDecryptedConfigs(
    baseUrl,
    apiKey,
    process.env.IDP_ENCRYPTION_KEY!,
    "my-app",
    "prod"
  );
  console.log(configs.DB_PASSWORD);

  const flags = await getFeatureFlags(baseUrl, apiKey, "my-app", "prod");
  if (flags.NEW_CHECKOUT) {
    console.log("NEW_CHECKOUT is enabled");
  }
}

if (require.main === module) {
  main();
}

File: examples/python/idp_client.py
"""
Example Python client for IDP: fetch every config/secret for a
service+environment (decrypted into a dict), and every feature flag
(into a dict of name -> enabled). Both use the same service client API key.

    pip install requests cryptography
    IDP_API_KEY=... IDP_ENCRYPTION_KEY=... python idp_client.py
"""
from __future__ import annotations

import os
from typing import Dict

import requests
from cryptography.fernet import Fernet


def get_decrypted_configs(
    base_url: str,
    api_key: str,
    encryption_key: str,
    service: str,
    environment: str,
) -> Dict[str, str]:
    """Fetch every config/secret for a service+environment and decrypt it into a dict."""
    response = requests.get(
        f"{base_url}/api/v1/config/configs/list",
        params={"service": service, "environment": environment},
        headers={"X-API-Key": api_key},
        timeout=10,
    )
    response.raise_for_status()

    fernet = Fernet(encryption_key.encode())
    return {
        entry["key"]: fernet.decrypt(entry["value"].encode()).decode()
        for entry in response.json()
    }


def get_feature_flags(
    base_url: str,
    api_key: str,
    service: str,
    environment: str,
) -> Dict[str, bool]:
    """Fetch every feature flag for a service+environment into a dict of name -> enabled."""
    response = requests.get(
        f"{base_url}/api/v1/config/feature-flags",
        params={"service": service, "environment": environment},
        headers={"X-API-Key": api_key},
        timeout=10,
    )
    response.raise_for_status()

    return {flag["name"]: flag["is_enabled"] for flag in response.json()}


if __name__ == "__main__":
    base_url = os.environ.get("IDP_BASE_URL", "http://localhost:8000")
    api_key = os.environ["IDP_API_KEY"]

    configs = get_decrypted_configs(
        base_url=base_url,
        api_key=api_key,
        encryption_key=os.environ["IDP_ENCRYPTION_KEY"],
        service="my-app",
        environment="prod",
    )
    print(configs.get("DB_PASSWORD"))

    flags = get_feature_flags(
        base_url=base_url,
        api_key=api_key,
        service="my-app",
        environment="prod",
    )
    if flags.get("NEW_CHECKOUT"):
        print("NEW_CHECKOUT is enabled")
